### PR TITLE
Enable golangci-lint on neonvm/

### PR DESCRIPTION
**Pull request by sharnoff**
Supersedes #400. With the recent PRs making changes to neonvm, it seemed like enabling golangci-lint would be more valuable. And with the amount of work required, I figured it'd be easier for someone more involved to tackle.

~~This PR builds on #720, but #718 could be used instead.~~

---

Notes for review:

Happy to clarify any weird bits. There's a lot here.

There were a few function signatures changed - either because they were returning a `(struct, error)` tuple (which trips exhaustruct - ends up beings simpler to return `(*struct, error)`), or because they had an unused parameter (or, newly unused, after switching some of the functions in `virtualmachine_controller.go` to use `*api.VCPUCgroup`).

Because of the size of #704, it may be worthwhile to get this merged before it. However, it's probably easier to resolve merge conflicts if that merges first. cc @omrigan for your thoughts.

### CODE REVIEW

File: .golangci.yml [hunk 1]
```diff
@@ -53,20 +50,30 @@ linters-settings:
...
+      - '^k8s\.io/client-go/tools/leaderelection/resourcelock\.ResourceLockConfig$'
+      - '^k8s\.io/client-go/tools/leaderelection\.(LeaderCallbacks|LeaderElectionConfig)$'
+      - '^sigs\.k8s\.io/controller-runtime/pkg/client\.Options$'
+      - '^sigs\.k8s\.io/controller-runtime/pkg/controller\.Options$'
+      - '^sigs\.k8s\.io/controller-runtime/pkg/envtest\.(Environment|WebhookInstallOptions)$'
+      - '^sigs\.k8s\.io/controller-runtime/pkg/manager\.Options$'
+      - '^sigs\.k8s\.io/controller-runtime/pkg/reconcile\.Result$'
+      - '^sigs\.k8s\.io/controller-runtime/pkg/scheme\.Builder$'
       - '^github.com/prometheus/client_golang/prometheus(/.*)?\.\w+Opts$'
-      - '^github\.com/containerd/cgroups/v3/cgroup2\.(Resources|Memory)'
```
Comment (sharnoff) [inline .golangci.yml:63]: Now unused. Plan to extract into a separate PR.

File: .golangci.yml [hunk 2]
```diff
@@ -109,18 +114,6 @@ linters-settings:
...
-      # Kubernetes' klog — this is our primary logger
-      - (k8s.io/klog/v2).Infof
-      - (k8s.io/klog/v2).Warningf
-      - (k8s.io/klog/v2).Errorf
-      - (k8s.io/klog/v2).Fatalf
-      # autoscaler-agent's RunnerLogger
-      - (github.com/neondatabase/autoscaling/pkg/agent.RunnerLogger).Infof
-      - (github.com/neondatabase/autoscaling/pkg/agent.RunnerLogger).Warningf
-      - (github.com/neondatabase/autoscaling/pkg/agent.RunnerLogger).Errorf
-      - (github.com/neondatabase/autoscaling/pkg/agent.RunnerLogger).Fatalf
```
Comment (sharnoff) [inline .golangci.yml:123]: Now unused. Plan to extract into a separate PR.

**Comment by sharnoff** [comment]
Hm, looks like lack of git worktrees support strikes again... :/
More lints to fix that didn't initially show up locally.

File: neonvm/controllers/virtualmachine_controller.go [hunk 3]
```diff
@@ -194,7 +190,7 @@ func (r *VirtualMachineReconciler) Reconcile(ctx context.Context, req ctrl.Reque
 }
 
 // finalizeVirtualMachine will perform the required operations before delete the CR.
```
Comment (shayanh) [inline neonvm/controllers/virtualmachine_controller.go:192]:
```suggestion
// doFinalizeVirtualMachine will perform the required operations before delete the CR.
```

File: neonvm/controllers/virtualmachine_controller.go [hunk 4]
```diff
@@ -194,7 +190,7 @@ func (r *VirtualMachineReconciler) Reconcile(ctx context.Context, req ctrl.Reque
...
-func (r *VirtualMachineReconciler) doFinalizerOperationsForVirtualMachine(ctx context.Context, virtualmachine *vmv1.VirtualMachine) error {
```
Comment (shayanh) [inline neonvm/controllers/virtualmachine_controller.go:197]: why do we ignore all the errors here?

**Review by shayanh — APPROVED** [review]
Looks good to me! Added a question and a small suggestion.

File: neonvm/controllers/virtualmachine_controller.go (hunk 3 above)
Comment (sharnoff) [reply neonvm/controllers/virtualmachine_controller.go:192]: should be `doFinalizerOperationsForVirtualMachine`, no?

File: neonvm/controllers/virtualmachine_controller.go (hunk 4 above)
Comment (sharnoff) [reply neonvm/controllers/virtualmachine_controller.go:197]: I don't know, honestly. AFAICT, this _probably_ means that it's possible to leak overlay IP addresses due to transient errors.
cc @cicdteam ?

File: neonvm/controllers/virtualmachine_controller.go (hunk 3 above)
Comment (shayanh) [reply neonvm/controllers/virtualmachine_controller.go:192]: ah right, my bad

**Review by Omrigan — APPROVED** [review]

**Comment by sharnoff** [comment]
I will merge this in 24h if no objections — want to get this in so PRs can build on it.
cc @cicdteam, @bayandin

**Review by bayandin — APPROVED** [review]
